invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,317 @@
+        if any(m.to_version == migration.to_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration to {migration.to_version} already registered")
+        self._migrations.add(migration)
+
+    def get(self, from_version: int) -> Optional[Migration]:
+        """Gets the migration that may be run on the given database version."""
+        # register() ensures that there is only one migration with a given from_version, so this is safe.
+        return next((m for m in self._migrations if m.from_version == from_version), None)
+
+    def validate_migration_chain(self) -> None:
+        """
+        Validates that the migrations form a single chain of migrations from version 0 to the latest version.
+        Raises a MigrationError if there is a problem.
+        """
+        if self.count == 0:
+            return
+        if self.latest_version == 0:
+            return
+        next_migration = self.get(from_version=0)
+        if next_migration is None:
+            raise MigrationError("Migration chain is fragmented")
+        touched_count = 1
+        while next_migration is not None:
+            next_migration = self.get(next_migration.to_version)
+            if next_migration is not None:
+                touched_count += 1
+        if touched_count != self.count:
+            raise MigrationError("Migration chain is fragmented")
+
+    @property
+    def count(self) -> int:
+        """The count of registered migrations."""
+        return len(self._migrations)
+
+    @property
+    def latest_version(self) -> int:
+        """Gets latest to_version among registered migrations. Returns 0 if there are no migrations registered."""
+        if self.count == 0:
+            return 0
+        return sorted(self._migrations, key=lambda m: m.to_version)[-1].to_version
+
+
+class SQLiteMigrator:
+    """
+    Manages migrations for a SQLite database.
+
+    :param db_path: The path to the database to migrate, or None if using an in-memory database.
+    :param conn: The connection to the database.
+    :param lock: A lock to use when running migrations.
+    :param logger: A logger to use for logging.
+    :param log_sql: Whether to log SQL statements. Only used when the log level is set to debug.
+
+    Migrations should be registered with :meth:`register_migration`.
+
+    During migration, a copy of the current database is made and the migrations are run on the copy. If the migration
+    is successful, the original database is backed up and the migrated database is moved to the original database's
+    path. If the migration fails, the original database is left untouched and the migrated database is deleted.
+
+    If the database is in-memory, no backup is made; the migration is run in-place.
+    """
+
+    backup_path: Optional[Path] = None
+
+    def __init__(
+        self,
+        db_path: Path | None,
+        conn: sqlite3.Connection,
```

Comment:
To support in-memory databases, we need to share a connection for the lifetime of the application, so this must be passed in.